fix: Negative idle cost on promless mode

spa-raj opened the pull request:
## What does this PR change?
 **Problem:**
  In promless mode, idle costs were going negative immediately after OpenCost startup.

  **Root Cause:**
  Pricing asymmetry between allocations and assets when collector metrics returned 0:
  - **Allocations**: Had fallback logic (allocation_helpers.go:1899) - when collector pricing = 0, use custom pricing
  - **Assets**: No fallback - kept $0 pricing when collector metrics unavailable
  - **Result**: `Idle = $0 (nodes) - $positive (pods) = NEGATIVE`

  **Fix:**
  Added same fallback logic to asset pricing in `cluster_helpers.go`:
  - `buildCPUCostMap()` and `buildRAMCostMap()` now check if collector returns 0 or NaN
  - When true, fall back to custom pricing (same as allocations)
  - Ensures pricing symmetry: both use same pricing source

  Files changed:
  - `pkg/costmodel/cluster_helpers.go` - Added fallback logic + math import
  - `pkg/costmodel/cluster_helpers_test.go` - Added test coverage

## Does this PR relate to any other PRs?
* No

## How will this PR impact users?
- Idle costs will now be **positive or zero** (correct) instead of negative (broken)
- Most noticeable immediately after OpenCost startup or when collector metrics are temporarily unavailable
- With debug logging enabled, users may see new log messages: `"ClusterNodes: node X has 0/NaN CPU cost from data source; using custom pricing fallback"`

## Does this PR address any GitHub or Zendesk issues?
* Fixes #3387 

## How was this PR tested?
- Added unit test case simulating collector returning 0
- All existing tests pass
- Verified idle costs are positive after OpenCost restart in promless mode

## Does this PR require changes to documentation?
* No

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
* This PR should be labled for next release but I don't have the access to label it.

spa-raj committed:
Fix negative idle costs in promless mode (#3387)

 When running OpenCost in promless mode with collector data source, idle
 costs could become negative due to inconsistent pricing fallback logic
 between allocations and assets.

 Root cause:
 - Allocations: When collector pricing metrics returned 0, fallback to
   custom pricing was applied (allocation_helpers.go:1899)
 - Assets: No fallback logic existed, so nodes kept $0 pricing when
   collector metrics were unavailable
 - Result: Idle = $0 (nodes) - $positive (allocations) = NEGATIVE

Solution:
 Added custom pricing fallback to buildCPUCostMap() and buildRAMCostMap()
 in cluster_helpers.go. When collector returns 0 or NaN for pricing
 metrics, both asset and allocation calculations now use the same custom
 pricing fallback, ensuring consistent costs and preventing negative idle.
 This fix mirrors the existing allocation fallback logic (checking for
 0 or NaN) on the asset side, maintaining pricing symmetry between the
 two calculation paths.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Add test coverage for asset pricing fallback in promless mode

Add test case to verify custom pricing fallback behavior when collector
data source returns 0 for node pricing metrics. This scenario occurs in
promless mode when collector hasn't scraped pricing data yet.

Changes:
 - Add "Collector returns 0, fallback to custom pricing" test case to
   TestAssetCustompricing
 - Add zeroCollector boolean flag to test case struct to simulate
   collector returning 0 values
 - Verify CPU and RAM pricing correctly falls back to custom pricing
   when data source returns 0 or NaN
 - Remove unused nodePromResult variable

This test ensures the asset pricing fallback logic mirrors the existing
allocation fallback behavior, preventing negative idle costs in promless
mode.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj commented:
## Note

  I was **unable to reproduce the negative idle cost bug** in my local environment because the collector was populating pricing metrics too quickly (within 1-2 seconds of startup).
  The timing window where collector returns 0 is very narrow.

  However, I verified the fix is correct through:

  * **Unit tests** - Added test case simulating collector returning 0, fallback logic triggers correctly
  * **Code analysis** - Traced both allocation and asset pricing paths, confirmed asymmetry was the root cause
  * **Integration testing** - Restarted OpenCost multiple times, idle costs remained positive (never negative)
  * **Logic validation** - The fix mirrors existing allocation fallback logic (allocation_helpers.go:1899)

  The fallback **will** trigger in production scenarios where:
  - Collector takes longer to scrape pricing data
  - Temporary collector failures occur  
  - Network latency delays metric availability
  - Initial OpenCost startup before collector has any data

ameijer committed:
fix window to be inclusive (#3395)

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

ameijer commented on pkg/costmodel/cluster_helpers.go:
```diff
@@ -88,8 +93,25 @@ func buildCPUCostMap(
+					cpuCost = customCPUCost
+					log.Debugf("ClusterNodes: node %s has 0/NaN CPU cost from data source; using custom pricing fallback: %f", name, cpuCost)
+				}
+			}
```
@spa-raj , I think I agree with your general approach here. I think though maybe you could just add this condition to the above if statement? 

(customPricingEnabled && customPricingConfig != nil ) || cpuCost == 0 || math.IsNaN(cpuCost)? Then, let's add a bunch of dedupedInfof logs here explaining that if the CPUCost was or or Nan, we log that it was invalid value, falling back to customer pricing, etc? 

I just feel like that would read a little easier

spa-raj committed:
Refactor custom pricing logic and add deduped info logs

Combine conditional checks for custom pricing into a single statement to
improve readability. Add DedupedInfof logs to explain why custom pricing
is being used (enabled, invalid 0/NaN from data source).

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Trigger PR update

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Trigger PR update

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Merge branch 'opencost:develop' into bug/negative-idle-cost-promless

ameijer commented on pkg/costmodel/cluster_helpers.go:
```diff
@@ -73,23 +74,34 @@ func buildCPUCostMap(
+						log.DedupedInfof(10, "ClusterNodes: node %s using custom pricing: %f", name, customCPUCost)
+					}
+					cpuCost = customCPUCost
+				}
```
if CPUCOst == 0 or CPU.IsNan { log "cpu was 0 or NAN, but was unable to fall back to custom pricing because it was nil"

spa-raj committed:
Warn when unable to fall back to custom pricing for invalid CPU costs

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

ameijer commented on pkg/costmodel/cluster_helpers.go:
```diff
@@ -141,23 +158,34 @@ func buildRAMCostMap(
+						log.DedupedInfof(10, "ClusterNodes: node %s using custom pricing: %f", name, customRAMCost)
+					}
+					ramCost = customRAMCost
+				}
```
can you add an else here too?

spa-raj committed:
Merge branch 'opencost:develop' into bug/negative-idle-cost-promless

spa-raj committed:
Add missing 'ram' fallback branch

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Merge branch 'opencost:develop' into bug/negative-idle-cost-promless

ameijer committed:
Merge branch 'develop' into bug/negative-idle-cost-promless

spa-raj committed:
Merge branch 'opencost:develop' into bug/negative-idle-cost-promless

spa-raj committed:
Merge branch 'opencost:develop' into bug/negative-idle-cost-promless

spa-raj committed:
Merge branch 'opencost:develop' into bug/negative-idle-cost-promless

spa-raj committed:
Merge branch 'opencost:develop' into bug/negative-idle-cost-promless

ameijer committed:
Merge branch 'develop' into bug/negative-idle-cost-promless

spa-raj committed:
fix: prevent negative idle costs in promless mode

Fixes an issue where idle allocations could have negative costs when
running in promless environments without custom pricing configured.

The bug occurred when:
 1. Node pricing data sources returned zero costs (promless mode)
 2. Cloud provider billing applied negative adjustments (discounts/credits)
 3. Asset totals became negative due to adjustments on zero base costs
 4. Idle calculation (assetTotal - allocTotal) resulted in negative values

This fix clamps idle costs (CPU, GPU, RAM) to a minimum of zero while
preserving all node data. Warning logs are emitted when clamping occurs
to provide visibility into pricing configuration issues.

The root cause should be addressed by ensuring custom pricing fallback
is configured in promless environments, but this defensive fix prevents
the user-facing issue of negative idle costs in allocation reports.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj commented:
@ameijer  Can you review the changes and also update the branch to publish new image to test my changes?

ameijer committed:
Merge branch 'develop' into bug/negative-idle-cost-promless

ameijer commented on pkg/costmodel/costmodel.go:
```diff
@@ -1788,20 +1817,45 @@ func computeIdleAllocations(allocSet *opencost.AllocationSet, assetSet *opencost
 		// asset cost and the allocated cost per-resource.
+		// Idle costs are clamped to zero to prevent negative values that can occur
+		// when asset costs are zero (e.g., in promless mode without custom pricing)
+		// but negative adjustments are applied from cloud providers.
```
this is interesting. you have adjustments, even though you have no cloud provider configured?

spa-raj committed:
fix: prevent negative idle costs by clamping to zero

Fixes an issue where idle allocations could have negative costs when
asset total costs are less than allocated costs.

This can occur in multiple scenarios:
 - Pricing data unavailable (promless mode, API failures, missing pricing)
 - Custom pricing misconfigured or returning zero values
 - Cloud billing adjustments reducing asset costs below allocations
 - Allocation calculations exceeding asset costs due to timing/rounding

The fix clamps idle costs (CPU, GPU, RAM) to max(0, assetCost - allocCost)
to enforce the mathematical invariant that idle costs cannot be negative.
Warning logs are emitted when clamping occurs to provide visibility into
the underlying pricing issues.

This is a defensive fix that prevents the user-facing bug while preserving
all node and allocation data for debugging. The root causes (pricing failures,
zero base costs with adjustments, etc.) should still be investigated and
addressed through proper pricing configuration.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj commented on pkg/costmodel/costmodel.go:
```diff
@@ -1817,7 +1817,35 @@ func computeIdleAllocations(allocSet *opencost.AllocationSet, assetSet *opencost
 		// Insert one idle allocation for each key (whether by node or
 		// by cluster), defined as the difference between the total
 		// asset cost and the allocated cost per-resource.
+		// Idle costs are clamped to zero to prevent negative values that can occur
+		// when asset total costs are less than allocated costs. This can happen when:
+		// - Pricing data is unavailable (promless mode, API failures, missing price data)
+		// - Custom pricing is misconfigured or returns zero values
+		// - Cloud billing adjustments reduce asset costs below allocation costs
+		// - Allocation calculations exceed asset costs due to timing or rounding
```
@ameijer Updated my comment here

spa-raj reviewed (commented):
This approach is working. Based on my investigation 
- Negative idle costs occur when asset total costs are less than allocated costs, which happens when asset pricing (`cluster_helpers.go`) and allocation pricing (`allocation_helpers.go`) use different fallback mechanisms in promless/pricing failure scenarios. The mathematical fix is to clamp idle costs to `max(0, assetCost - allocCost)` in computeIdleAllocations() since idle resources cannot logically have negative cost.

<img width="2558" height="1328" alt="image" src="https://github.com/user-attachments/assets/6a8c3f02-51ff-4e5c-b798-e95c99581310" />

ameijer commented on pkg/costmodel/costmodel.go:
```diff
@@ -1817,7 +1817,35 @@ func computeIdleAllocations(allocSet *opencost.AllocationSet, assetSet *opencost
+
+		// Clamp idle costs to zero to prevent negative idle allocations
+		if cpuIdleCost < 0 {
+			log.DedupedWarningf(5, "Allocation: clamping negative CPU idle cost to 0 for %s (assetTotal: %.4f, allocTotal: %.4f)",
```
make un deduped

ameijer commented on pkg/costmodel/costmodel.go:
```diff
@@ -1817,7 +1817,35 @@ func computeIdleAllocations(allocSet *opencost.AllocationSet, assetSet *opencost
+
+		cpuIdleCost := assetTotal.TotalCPUCost() - allocTotal.TotalCPUCost()
+		gpuIdleCost := assetTotal.TotalGPUCost() - allocTotal.TotalGPUCost()
+		ramIdleCost := assetTotal.TotalRAMCost() - allocTotal.TotalRAMCost()
```
let's add debug functionality here that takes the assets and allocation sets. it should compute 
1. any allocations without matching asset (match via providerID) 
2. Sum total of allocations for each node asset, compare to the node asset
3. sum total of node costs
4. sum total of allocation costs

ameijer replied:
might want to also add resource amounts to totals structs, then we can compare amounts vs costs (want higher total resources to have higher costs)

spa-raj committed:
Add resource amount tracking to allocation and asset totals for pricing validation

Add CPUCoreHours, GPUHours, and RAMByteHours fields to AllocationTotals
and AssetTotals structs to enable comparison of resource amounts vs costs.
This helps identify pricing inversions where higher resource usage doesn't
correlate with higher costs, which can indicate pricing data issues.

Changes:
 - Add resource hour fields (CPUCoreHours, GPUHours, RAMByteHours) to
   AllocationTotals and AssetTotals structs
 - Update ComputeAllocationTotals and ComputeAssetTotals to accumulate
   resource amounts
 - Update Clone methods to include new fields
 - Enhance debugAssetAllocationMismatch to detect and warn about
   resource-cost inversions per node
 - Add debug logging of resource hours alongside costs for comparison
 - Change negative idle cost logs from deduplicated to standard warnings
   with clearer messaging

This enables detection of scenarios where:
 - Assets have more resource hours but lower costs than allocations
 - Assets have fewer resource hours but higher costs than allocations
 Both indicate potential pricing data issues in promless mode or with
 custom pricing configurations.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj commented:
@ameijer @nikovacevic @Sean-Holcomb I have made the changes we discussed in our call. Please take a look.

ameijer committed:
Merge branch 'develop' into bug/negative-idle-cost-promless
